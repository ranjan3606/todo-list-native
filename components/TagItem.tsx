import React, { useRef } from 'react';
import {
  View,
  Text,
  StyleSheet,
  TouchableOpacity,
  Animated,
  Pressable,
  Alert
} from 'react-native';
import { FontAwesome } from '@expo/vector-icons';
import { Colors } from '@/constants/Colors';
import { getTags, saveTags } from '@/services/storage';
import { todoEvents, TODO_EVENTS } from '@/services/eventEmitter';
import { useTranslation } from '@/i18n';

interface TagItemProps {
  tagName: string;
  keywords: string[];
  isActive: boolean;
  colorScheme: 'light' | 'dark';
  animValue?: Animated.Value; // Make animValue optional
  onPress: () => void;
  onEdit: () => void;
}

export const TagItem = ({
  tagName,
  keywords,
  isActive,
  colorScheme,
  animValue,
  onPress,
  onEdit
}: TagItemProps) => {
  const { t } = useTranslation();
  
  // Create a default animation value if none is provided
  const defaultAnimValue = useRef(new Animated.Value(100)).current;
  const safeAnimValue = animValue || defaultAnimValue;
  
  const handleDelete = async () => {
    Alert.alert(
      t('tags.deleteTag'),
      t('tags.deleteConfirm', { name: tagName }),
      [
        { text: t('common.cancel'), style: 'cancel' },
        {
          text: t('common.delete'),
          style: 'destructive',
          onPress: async () => {
            try {
              // Get current tags
              const currentTags = await getTags();
              // Create a new object without the deleted tag
              const { [tagName]: removed, ...updatedTags } = currentTags;
              
              // Save updated tags
              await saveTags(updatedTags);
              
              // Emit tag deleted event
              todoEvents.emit(TODO_EVENTS.TAG_DELETED, { name: tagName });
            } catch (error) {
              console.error('Error deleting tag:', error);
              Alert.alert(t('common.error'), t('tags.deleteError'));
            }
          }
        }
      ]
    );
  };
  
  return (
    <View style={[styles.tagItem, { backgroundColor: Colors[colorScheme].cardBackground }]}>
      <Pressable 
        onPress={onPress}
        android_ripple={{ color: 'rgba(0,0,0,0.1)' }}
        style={({ pressed }) => [
          styles.tagInfo,
          pressed && { backgroundColor: 'rgba(0,0,0,0.05)' }
        ]}
      >
        <View style={styles.tagContent}>
          <Text style={[styles.tagName, { color: Colors[colorScheme].text }]}>
            {tagName}
          </Text>
          <Text style={[styles.tagKeywords, { color: Colors[colorScheme].lightText }]}>
            {keywords.join(', ')}
          </Text>
        </View>
        
        <Animated.View 
          style={[
            styles.animatedTagActions,
            { 
              transform: [{ translateX: safeAnimValue }] 
            }
          ]}
        >
          <TouchableOpacity 
            onPress={onEdit} 
            style={styles.actionButton}
            testID="edit-tag-button"
            accessibilityRole="button"
          >
            <FontAwesome name="pencil" size={18} color={Colors.primary} />
          </TouchableOpacity>
          <TouchableOpacity 
            onPress={handleDelete} 
            style={styles.actionButton}
            testID="delete-tag-button"
            accessibilityRole="button"
          >
            <FontAwesome name="trash" size={18} color="red" />
          </TouchableOpacity>
        </Animated.View>
      </Pressable>
    </View>
  );
};

const styles = StyleSheet.create({
  tagItem: {
    borderRadius: 8,
    marginBottom: 10,
    padding: 12,
  },
  tagInfo: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    overflow: 'hidden',
    position: 'relative',
    borderRadius: 8,
  },
  tagContent: {
    flex: 1,
    padding: 5,
  },
  animatedTagActions: {
    flexDirection: 'row',
    alignItems: 'center',
    position: 'absolute',
    right: 0,
    top: 0,
    justifyContent: 'center',
    paddingHorizontal: 5,
  },
  tagName: {
    fontSize: 16,
    fontWeight: '600',
    marginBottom: 4,
  },
  tagKeywords: {
    fontSize: 14,
  },
  actionButton: {
    padding: 8,
    marginLeft: 8,
  },
});
