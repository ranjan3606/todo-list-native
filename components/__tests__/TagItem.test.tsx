import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { TagItem } from '../TagItem';

// Mock expo-notifications to prevent deprecation warnings
jest.mock('expo-notifications', () => ({
  getExpoPushTokenAsync: jest.fn(() => Promise.resolve({ data: 'mock-push-token' })),
  setNotificationHandler: jest.fn(),
  scheduleNotificationAsync: jest.fn(() => Promise.resolve('notification-id')),
  dismissNotificationAsync: jest.fn(),
  cancelScheduledNotificationAsync: jest.fn(),
  getBadgeCountAsync: jest.fn(() => Promise.resolve(0)),
  setBadgeCountAsync: jest.fn(() => Promise.resolve(true)),
}));

jest.mock('@react-native-async-storage/async-storage', () => ({
  getItem: jest.fn(() => Promise.resolve(null)),
  setItem: jest.fn(() => Promise.resolve(null)),
  removeItem: jest.fn(() => Promise.resolve(null)),
  clear: jest.fn(() => Promise.resolve(null)),
  getAllKeys: jest.fn(() => Promise.resolve([])),
  multiGet: jest.fn(() => Promise.resolve([])),
  multiSet: jest.fn(() => Promise.resolve(null)),
  multiRemove: jest.fn(() => Promise.resolve(null)),
}));

jest.mock('@expo/vector-icons', () => {
  const React = require('react');
  const { View } = require('react-native');
  return {
    FontAwesome: function MockFontAwesome(props: any) {
      return React.createElement(View, {...props, testID: `icon-${props.name}`});
    }
  };
});

jest.mock('@/i18n', () => ({
  useTranslation: () => ({
    t: (key: string, params?: any) => params ? `${key}:${JSON.stringify(params)}` : key
  })
}));

jest.mock('@/constants/Colors', () => ({
  Colors: {
    light: { cardBackground: '#fff', text: '#000', lightText: '#888' },
    dark: { cardBackground: '#000', text: '#fff', lightText: '#ccc' },
    primary: '#007bff',
  }
}));

describe('TagItem', () => {
  const baseProps = {
    tagName: 'work',
    keywords: ['job', 'office'],
    isActive: false,
    colorScheme: 'light' as const,
    onPress: jest.fn(),
    onEdit: jest.fn(),
  };

  it('calls onPress when pressed', () => {
    const onPress = jest.fn();
    const { getByText } = render(<TagItem {...baseProps} onPress={onPress} />);
    fireEvent.press(getByText('work'));
    expect(onPress).toHaveBeenCalled();
  });

  it('calls onEdit when edit button is pressed', () => {
    const onEdit = jest.fn();
    const { getByTestId } = render(<TagItem {...baseProps} onEdit={onEdit} isActive={true} />);
    fireEvent.press(getByTestId('edit-tag-button'));
    expect(onEdit).toHaveBeenCalled();
  });
});
