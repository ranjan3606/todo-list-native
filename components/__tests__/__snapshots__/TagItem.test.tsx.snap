// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagItem renders correctly (snapshot) 1`] = `
<View
  style={
    [
      {
        "borderRadius": 8,
        "marginBottom": 10,
        "padding": 12,
      },
      {
        "backgroundColor": "#fff",
      },
    ]
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "borderRadius": 8,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "overflow": "hidden",
          "position": "relative",
        },
        false,
      ]
    }
  >
    <View
      style={
        {
          "flex": 1,
          "padding": 5,
        }
      }
    >
      <Text
        style={
          [
            {
              "fontSize": 16,
              "fontWeight": "600",
              "marginBottom": 4,
            },
            {
              "color": "#000",
            },
          ]
        }
      >
        work
      </Text>
      <Text
        style={
          [
            {
              "fontSize": 14,
            },
            {
              "color": "#888",
            },
          ]
        }
      >
        job, office
      </Text>
    </View>
    <View
      collapsable={false}
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
          "paddingHorizontal": 5,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "transform": [
            {
              "translateX": 100,
            },
          ],
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "marginLeft": 8,
            "opacity": 1,
            "padding": 8,
          }
        }
        testID="edit-tag-button"
      >
        <View
          color="#007bff"
          name="pencil"
          size={18}
          testID="icon-pencil"
        />
      </View>
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "marginLeft": 8,
            "opacity": 1,
            "padding": 8,
          }
        }
        testID="delete-tag-button"
      >
        <View
          color="red"
          name="trash"
          size={18}
          testID="icon-trash"
        />
      </View>
    </View>
  </View>
</View>
`;
